group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java

```diff
@@ -2904,6 +2908,135 @@ private void validateGenericGroupHeartbeat(
+                    );
+                } else {
+                    try {
+                        group.validateMember(member.memberId(), member.groupInstanceId(), "leave-group");
+                        coordinatorResult = removeCurrentMemberFromGenericGroup(
+                            group,
+                            member.memberId(),
+                            member.reason()
+                        );
+                        memberResponses.add(
+                            new MemberResponse()
+                                .setMemberId(member.memberId())
+                                .setGroupInstanceId(member.groupInstanceId())
+                        );
+                    } catch (KafkaException e) {
+                        memberResponses.add(
+                            new MemberResponse()
+                                .setMemberId(member.memberId())
+                                .setGroupInstanceId(member.groupInstanceId())
+                                .setErrorCode(Errors.forException(e).code())
+                        );
+                    }
+                }
+            }
+            return new CoordinatorResult<>(
+                coordinatorResult.records(),
+                new LeaveGroupResponseData()
+                    .setMembers(memberResponses),
+                coordinatorResult.appendFuture()
+            );
+    }
+
+    /**
+     * Remove a member from the group. Cancel member's heartbeat, and prepare rebalance
+     * or complete the join phase if necessary.
+     * 
+     * @param group     The generic group.
+     * @param memberId  The member id.
+     * @param reason    The reason for the LeaveGroup request.
+     *
+     * @return The GroupMetadata record and the append future to be completed once the record is
+     *         appended to the log (and replicated).
+     */
+    private CoordinatorResult<Void, Record> removeCurrentMemberFromGenericGroup(
+        GenericGroup group,
+        String memberId,
+        String reason
+    ) {
+        GenericGroupMember member = group.member(memberId);
+        reason = reason != null ? reason : "not provided";
+        timer.cancel(genericGroupHeartbeatKey(group.groupId(), memberId));
+        log.info("[Group {}] Member {} has left group through explicit `LeaveGroup` request; client reason: {}",
+            group.groupId(), memberId, reason);
+
+        group.completeJoinFuture(member,
+            new JoinGroupResponseData()
+                .setMemberId(UNKNOWN_MEMBER_ID)
+                .setErrorCode(Errors.UNKNOWN_MEMBER_ID.code())
+        );
+        group.remove(member.memberId());
+
+        switch (group.currentState()) {
+            case STABLE:
+            case COMPLETING_REBALANCE:
+                return maybePrepareRebalanceOrCompleteJoin(group, reason);
+            case PREPARING_REBALANCE:
+                timer.cancel(genericGroupJoinKey(group.groupId()));
```

Comment:
This does not seem correct to me. I think that we should cancel the timer only if we transition to the next state, no?

Reply:
Thanks for the catch. I updated this to `maybeCompleteJoinPhase(group)` but one difference is that the existing code (`GroupCoordinator#tryCompleteJoin`) completes the join phase even if it's the initial rebalance whereas `maybeCompleteJoinPhase` does not.

Do you think this will be an issue? In the case that it is the initial rebalance, then the timer will retry after the timeout.

Reply:
That makes sense to me.